File: LeetCode/0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        int n = heights.length , ans = 0, i = 0;
        Stack<Integer> st = new Stack<>();
        st.push(0);
        while (i < n) {
            while (!st.isEmpty() && heights[st.peek()] > heights[i]) {
                int index = st.peek();
                st.pop();
                int height = heights[index];
                
                if (st.isEmpty()) {
                    ans = Math.max(ans , height * i);
                } else {
                    int len = i - st.peek() - 1;
                    ans = Math.max(ans , height * len);
                }
            }
            st.push(i);
            i++;
        }
        while (!st.isEmpty()) {
            int index = st.peek();
            st.pop();
            int height = heights[index];
            
            if (st.isEmpty()) {
                ans = Math.max(ans , height * i);
            } else {
                int len = i - st.peek() - 1;
                ans = Math.max(ans , height * len);
            }
        }
        
        return ans;
    }
}

File: LeetCode/0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        n , i , ans = len(heights) , 0 , 0
        st = []
        
        while i < n:
            while st and heights[st[-1]] > heights[i]:
                t = st.pop()
                height = heights[t]
                
                if not st:
                    ans = max(ans , height * i)
                else:
                    length = i - st[-1] -1
                    ans = max(ans , height * length)
            st.append(i)
            i += 1
            
        while st:
            t = st.pop()
            height = heights[t]
            
            if not st:
                ans = max(ans , height * i)
            else:
                length = i - st[-1] -1
                ans = max(ans , height * length)
                
        return ans